---
title: "snRNA-seq using dataset integration"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
        df_print: paged
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE,
                      message=FALSE,
                      cache.lazy=FALSE)
```

This workflow is designed to perform single cell RNA-seq (scRNA-seq) using 
[scanpy](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0).

```{r rpackages}
library(reticulate)
library(tidyr)
library(stringr)
library(dplyr)
library(Seurat)
library(ggplot2)
library(cowplot)
library(DT)
library(plotly)
library(readr)
library(UpSetR)
library(future)
library(parallel)
library(scater)
library(Seurat)
library(scuttle)
library(purrr)
library(viridis)
source('../helpers.R')
```

```{r config}
# Path to Seurat obj
seurat.path <- '../output/scanpy-integrated-markers-all/seurat_integrated_markers.rds'

# Conda env used
conda.path <- '../../env'

# Path to sampletable
sampletable.path <- 'sampletable.tsv'

# Metadata column names for each clustering
clustering <- c(unintegrated='leiden_2.0', integrated='scvi_leiden_1.4')

# Path to output directory
out.path <- '../output/scanpy-to-seurat'

# Create the output directory if not present
if (!dir.exists(out.path)) { dir.create(out.path, recursive=TRUE) }
```

# Conda environment

```{r conda}
print(paste(conda.path, 'used...'))
```

# Importing Seurat obj

```{r import}
obj <- readRDS(seurat.path)
print(paste('Seurat object saved in', seurat.path, 'imported!'))
print(obj)
```

# Reduced dimensions

Reduced dimensions are explored using `obj@reductions`.

```{r available_dimensions}

print(obj@reductions)

```

# Reduced dimensions with clustering {.tabset}

```{r dimensions_cluster, results='asis', fig.width=12, fig.height=10}

for (reduction in names(obj@reductions)) {
    cat('##', reduction, '{.tabset}\n\n')
    for (cl in names(clustering)) {
        cat('###', cl, '\n\n')

            p <- DimPlot(
                obj,
                pt.size=0.1,
                reduction=reduction,
                group.by=clustering[cl],
                label=TRUE,
                repel=TRUE
                )
            print(p)
            cat('\n\n')
            }
            }


```

# Reduced dimensions with dataset {.tabset}


```{r dimensions_dataset, results='asis', width=14, fig.height=7}

for (reduction in names(obj@reductions)) {
    cat('##', reduction, '\n\n')
    p <- DimPlot(
        obj,
        pt.size=0.1,
        reduction=reduction,
        group.by='group'
        )
    print(p)
    cat('\n\n')
}


```

# Single cell metadata

Metadata is exported as an excel-readable tab-separated file with the following columns:

- **`barcodes`**: single cell barcodes
- **`barcodes_suffix`**: single cell barcodes with suffixes
- **`samplename`**: `<age>_<genotype>_<sex>_<rep>`
- **`group`**: `<age>_<genotype>_<sex>`
- **`leiden_2.0`**: unintegrated clustering
- **`scvi_leiden_1.4`**: integrated clustering

The suffix next to each barcode was added during downstream analysis to avoid confusion 
arising from identical barcodes in different datasets. All the Seurat objects provided 
in the current project use the barcodes with suffixes.

```{r barcodes, results='asis'}

# Import sampletable
sample.df <- read_tsv(sampletable.path)[, c('samplename', 'data_path')]

# Retrieve Seurat metadata as a data frame
meta.df <- obj@meta.data[, c('barcodes', 'samplename', 'group', unname(clustering))]

# Break if sample names don't match
if (!identical(sort(unique(meta.df$samplename)), sort(sample.df$samplename))) {
    stop("Sample names unmatched between the sample table and Seurat metadata!")
}

# Join two data frames
joined.df <- inner_join(
    meta.df,
    sample.df,
    by="samplename") %>%
    # Copy the barcode column to a new column named barcodes_suffix
    mutate(barcodes_suffix=barcodes)
# Strip the suffix in the barcodes column
joined.df[['barcodes']] <- map_chr(
    joined.df[['barcodes_suffix']],
    ~str_split_i(.x, pattern="-", i=1)
    )

# Create a vector where column names are reordered
joined.cols <- colnames(joined.df)
joined.cols <- c(
    joined.cols[str_detect(joined.cols, "barcodes")],
    joined.cols[!str_detect(joined.cols, "barcodes")]
)

# Reorder columns in the data frame
joined.df <- joined.df[, joined.cols]

# Save as a tsv file
tsv.path <- file.path(out.path, "barcode_table.tsv")
write.table(
    joined.df,
    tsv.path,
    quote=FALSE,
    row.names=FALSE,
    sep="\t"
)

link.table(tsv.path)
```

# Session info

```{r session_info, collapse=FALSE}
sessionInfo()
```
