---
title: "Marker gene computation on integrated dataset"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
        df_print: paged
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE,
                      message=FALSE,
                      cache.lazy=FALSE)
```

This workflow is designed to perform single cell RNA-seq (scRNA-seq) using 
[scanpy](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0).

```{r rpackages}
library(reticulate)
library(tidyr)
library(dplyr)
library(Seurat)
library(ggplot2)
library(cowplot)
library(DT)
library(plotly)
library(readr)
library(UpSetR)
library(future)
library(parallel)
library(SingleCellExperiment)
library(scater)
library(Seurat)
library(scuttle)
library(purrr)
library(anndata)
library(viridis)
library(Matrix)
source('../helpers.R')
use_condaenv('../../env')
```


```{python ppackages}
# Import packages
import numpy as np
import pandas as pd
import scanpy as sc
import seaborn as sns
import matplotlib.pyplot as plt
import anndata as ad
import scvi
import os
import random
# import torch
# Set seed number
scvi.settings.seed = 0
```

```{python environmental_variables}
os.environ['CUDA_LAUNCH_BLOCKING'] = '1'
os.environ['CUDA_VISIBLE_DEVICES'] = '1'
os.environ['TORCH_CUDNN_V8_API_DISABLED'] = '1'
```

```{r import_scanpy}
sc <- import("scanpy")
```

```{python configs}

# reticulate::repl_python()
# Assign path to input sampletable
input_sampletable = 'sampletable.tsv'

# Output directory
output_dir = '../output/scanpy-integrated-markers-all'

# Number of PCs to be accounted for UMAP
npcs = 50

# Assign resolution ranges for clustering
resolutions = [0.6, 0.8, 1, 1.2, 1.4, 1.6, 1.8, 2.0]

# Assign community detection algorithm
# Options:
# - leiden
# - louvain
comm_detection = 'leiden'

# Assign the number of top variable features
n_features = 3000

# Create the output directory if it's not present yet
if not os.path.exists(output_dir):
    os.makedirs(output_dir)

# Assign group levels
group_levels = [
    '5w_WT_M',
    '5w_cKO_M',
    '10w_WT_M',
    '10w_cKO_M'
    ]


# Assign input adata path
adata_path = {
    'unintegrated':'../output/scanpy-unintegrated-markers-all/adata_unintegrated_markers.h5ad',
    'integrated':'../output/scanpy-integration-all/adata_integrated.h5ad'
    }

# Assign a list storing names of metadata columns for marker gene groups
marker_groups = ['group', 'genotype', 'age', 'leiden_2.0', 'scvi_leiden_1.4']
cluster_col = 'scvi_leiden_1.4'
cluster_col_all = ['leiden_2.0', 'scvi_leiden_1.4']

# Set min threshold filtering marker genes
min_marker_l2fc = 0.25
min_marker_fraction = 0.1
alpha = 0.1 # FDR cutoff

# Assign genes of interest
genes = {
    'cholinergic':['Chat'],
    'visceral MNs':['Zeb2'],
    'Cholinergic interneurons':['Chat', 'Mpped2', 'Gad2'],
    'Preganglionic cells (PGCs)':['Chat', 'Fbn2'],
    'skeletal MNs':['Tns1'],
    'alpha':['Rbfox3', 'Stk32a', 'Sv2b'],
    'beta':['Gpr149'],
    'gamma':['Nrp2', 'Pard3b']
}
```

# Dataset

Previously created `adata` object is loaded. We expect this object 
to be QC-ed, dim-reduced, clustered, and integrated in the previous step.


```{python import_adata}

adata_dic = {}
# Load adata object
for key, value in adata_path.items():
    adata_dic[key] = sc.read_h5ad(value)

```

## Table {.tabset}

The number of cells are summarized in the following tables.

Tabs:

- **`samplename`**: cells per samplename
- **`group`**: cells per age per genotype
- **`genotype`**: cells per genotype
- **`age`**: cells per age
- **`leiden_2.0`**: cells clustered WITHOUT integration
- **`scvi_leiden_1.4`**: cells clustered WITH integration

```{r overview_table, results='asis'}

# Extract metadata from the `adata` and slice columns of interest
meta <- py$adata_dic[['integrated']]$obs[, c('samplename', py$marker_groups)]

# > head(meta, 2)
#                    samplename group leiden_2.0
# AAACCCAAGAGCTGCA-1  6w_WT_F_1 6w_WT         42
# AAACCCAAGCTAATGA-1  6w_WT_F_1 6w_WT         13


# Loop over to display the number of cells across the categories in each column
for (name in colnames(meta)) {
    cat('###', name, '\n\n')
    df <- as.data.frame(table(meta[[name]]))
    colnames(df) <- c('Category', 'nCell')
    subchunkify(paste0(name, '_overview_table'))
    cat('\n\n')
}
```

## Plot {.tabset}


```{r overview_plot, results='asis'}

for (name in colnames(meta)) {

    cat('###', name, '\n\n')

    # Extract and clean input data frame for metadata
    df <- meta %>%
        group_by_at(vars(name)) %>%
        summarize(nCell=n())

    # Plot
    p <- ggplot(df, aes_string(x=name, y='nCell', fill=name)) +
        geom_bar(stat='identity', width=0.5) +
        theme_bw() +
        theme(legend.title=element_blank(),
              legend.position="none",
              axis.text.x=element_text(angle=90, vjust=0.5, hjust=1))

    subchunkify(paste0(name, '_overview_plot'),
                input='plot', 10, 6, ggplotly=FALSE)
    cat('\n\n')
}


```



# Marker genes

Here we calculate marker genes per cluster and per group. 
By default, itâ€™s calculated for one cluster vs the rest.

```{r notification_and_rebuild_obj}
# NOTE:
# We cannot use adata from integrated one since it only has N highly variable genes.
# Therefore, we combine two adatas:
# - unintegrated adata for log-normalized read count matrix
# - UMAP coordinates from integrated adata
# - metadata from both unintegrated and integrated adatas

# Build a new adata from unintegrated one
py$adata <- py$adata_dic[['unintegrated']]
integrated_adata <- py$adata_dic[['integrated']]

# Update metadata
py$adata$obs <- integrated_adata$obs

print("One vs the rest will be compared in the following categories:")

for (g in py$marker_groups) {
    print(paste0('- ', g, ': '))
    meta <- py$adata$obs
    print(levels(meta[[g]]))
    cat('\n')
}

# Replace umap coordinates with integrated ones
py$adata$obsm$X_umap <- integrated_adata$obsm$X_umap

# Add X_scVI
py$adata$obsm$X_scVI <- integrated_adata$obsm$X_scVI

# Add unintegrated umap coordinates
py$adata$obsm$X_unintegrated_umap <- py$adata_dic[['integrated']]$obsm$X_unintegrated_umap

```


```{python marker_gene_default}

# Function references:
# - https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.rank_genes_groups.html
# - https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.filter_rank_genes_groups.html

for g in marker_groups:
    # Compute marker genes
    sc.tl.rank_genes_groups(
        adata,
        groupby=g,   # metadata variable (e.g. clustering or samplegroup)
        key_added=f"integrated_{g}",   # name of adata.uns storing marker genes
        method='wilcoxon',
        bts=True)   # computes fraction of cells expressing each gene
```

```{python marker_genes_subset, cache=TRUE}

# Prep a list having clusters increasingly ordered
reordered_clusters = sorted(list(set(adata.obs.loc[:, cluster_col])), key=int)

# Create a dictionary storing subsetted adata
adata_sub_dic = {}
for c in reordered_clusters:
    idx = adata.obs[cluster_col] == c
    adata_sub = adata[idx, :].copy()
    adata_sub_dic[c] = adata_sub


valid_subset_keys = []
subset_groups = marker_groups[:-2]

failed_keys = []

for g in subset_groups:
    for c in adata_sub_dic.keys():
        this_key = f"integrated_{g}_in_cluster{c}"
        try:
            # compute marker genes
            sc.tl.rank_genes_groups(
                adata_sub_dic[c],  # takes subsetted AnnData as input obj
                groupby=g,   # metadata variable (e.g. clustering or samplegroup)
                key_added=this_key,   # name of adata.uns storing marker genes
                method='wilcoxon',
                bts=True)   # computes fraction of cells expressing each gene
            valid_subset_keys.append(this_key)
        except (ValueError, ZeroDivisionError):
            failed_keys.append(this_key)

if len(failed_keys) > 0:
    print("No pairs found in the following clusters:")
    for k in failed_keys:
        print(k)

```



## Tables (all cells) {.tabset}

Marker genes are filtered based on the following thresholds

- Absolute `logfoldchange` >= `r py$min_marker_l2fc`
- False discovery rate (FDR) `pvals_adj` <= `r py$alpha`

Due to file size limit, 2000 most top FDR genes are displayed 
in the following tables. Download the attached excel-readable files 
for total marker genes.

Clusters having no marker genes are omitted.

Table columns:

- **`names`**: gene symbols
- **`scores`**: z-scores underlying the computation of a p-value 
  for each gene for each group
- **`pvals`**: raw p-values
- **`pvals_adj`**: FDR computed by 
  [BH method](https://link.springer.com/referenceworkentry/10.1007/978-1-4419-9863-7_1215)
- **`cluster`**: a cluster in which each gene is differentially expressed 
  compared to the rest of the clusters
- **`nclusters`**: the number of clusters expressing each gene 
  as a differentially expressed gene (DEG)

The _z-score_ is a statistical metric indicating how many standard deviations 
a data point is from the mean of a group of values. If a z-score is zero, 
the datapoint is identical to the mean. A z-score of 1.0 would indicate a value 
that is one standard deviation away from the mean.

We separate results into **all markers** and **unique markers**, as described below:

- **All markers**: This is the full set of results, where one gene can 
  be a marker for multiple clusters.
- **Unique markers**: This only contains genes that were found as a marker 
  in a single cluster.


```{r marker_df, results='asis'}

# NOTE:
# You run sc$get$rank_genes_groups_df to extract marker genes as a data frame in R.
# It corresponds to `sc.tl.rank_genes_group()` in python.

marker.list <- list()

for (key in py$marker_groups) {

    cat('###', key, '{.tabset}\n\n')
    # Create a data frame storing all markers
    df <- mclapply(levels(py$adata$obs[[key]]), function(x)
        sc$get$rank_genes_groups_df(
            py$adata,
            key=paste0('integrated_', key),   # same as `key_added` in the `sc.tl.rank_genes_groups`
            group=x    # same as the `groupby` in the `sc.tl.rank_genes_groups`
            ) %>%
        mutate(cluster=x)
        ) %>% 
    bind_rows() %>%
    # Remove unwanted genes having FDR below threshold AND
    # above logfoldchange threshold
    dplyr::filter(pvals_adj <= py$alpha,
                  abs(logfoldchanges) >= py$min_marker_l2fc)

    # Add a column indicating the number of occurrences of the marker genes
    cluster.df <- df %>%
        group_by(names) %>%
        summarize(nclusters=n())
    df <- left_join(df, cluster.df, by='names') %>%
        unique()

    # Add the data frame to `marker.list`
    marker.list[[key]] <- df

    # Save total marker genes
    file.i <- file.path(py$output_dir,
                        paste0(key, '_all_markers.tsv'))
    write.table(df,
                file.i,
                quote=FALSE,
                sep='\t',
                row.names=FALSE,
                col.names=TRUE)

    cat('#### All markers {.tabset}\n\n')
    link.table(file.i)
    # Loop over to print 2000 top FDR genes in each cluster
    # Split the data frame by cluster
    df.list <- split(df, df$cluster)
    # Reorder the list
    df.list <- df.list[levels(py$adata$obs[[key]])]
    # Remove NULL if present
    df.list <- df.list[!is.na(names(df.list))]
    for (name in names(df.list)) {
        cat('#####', name, '\n\n')
        # Subset top 2000 FDR genes
        df <- df.list[[name]] %>%
            arrange(pvals_adj) %>%
            head(2000)
        # Remove rownames
        rownames(df) <- NULL
        subchunkify(paste0(key, '_', name, '_allmarkers'))
        cat('\n\n')
    }

    cat('#### Unique markers {.tabset}\n\n')

    # Clean the data frame to remove non-unique genes
    df <- do.call('rbind', df.list) %>%
        dplyr::filter(nclusters == 1)
    # Save total marker genes
    file.i <- file.path(py$output_dir,
                        paste0(key, '_unique_markers.tsv'))
    write.table(df,
                file.i,
                quote=FALSE,
                sep='\t',
                row.names=FALSE,
                col.names=TRUE)

    link.table(file.i)

    # Loop over to print 2000 top FDR genes in each cluster
    # Split the data frame by cluster
    df.list <- split(df, df$cluster)
    # Reorder the list
    df.list <- df.list[levels(py$adata$obs[[key]])]
    # Remove NULL if present
    df.list <- df.list[!is.na(names(df.list))]
    for (name in names(df.list)) {
        cat('#####', name, '\n\n')
        # Subset top 2000 FDR genes
        df <- df.list[[name]] %>%
            arrange(pvals_adj) %>%
            head(2000)
        # Remove rownames
        rownames(df) <- NULL
        subchunkify(paste0(key, '_', name, '_uniquemarkers'))
        cat('\n\n')
    }
}

```

```{r marker_df_subset, cache=TRUE}

# Retrieve the dictionary of adata subsets to a list
adata_sub_list <- py$adata_sub_dic

# Create an empty
subset_marker_list <- list()

# Loop over experimental conditions (e.g. genotype)
for (condition in py$subset_groups) {
    subset_marker_list[[condition]] <- list()
    # Loop over clusters
    for (name in names(adata_sub_list)) {
        adata_sub <- adata_sub_list[[name]]
        # Create a vector for condition levels (e.g. WT cKO)
        all_subsets <- levels(adata_sub$obs[[condition]])

        # Define the key used to store marker genes in the previous step
        this_key <- paste0(
            "integrated_", 
            condition,
            "_in_cluster",
            name
        )

        # If the key was valid for DE testing
        if (this_key %in% py$valid_subset_keys) {

            # Extract data frames
            df <- mclapply(all_subsets, function(x) {
                sc$get$rank_genes_groups_df(
                    adata_sub,
                    key=this_key,
                    group=x
                    ) %>%
                    # Add extra info
                    mutate(cluster=x) %>%
                    # Remove unwanted genes having FDR below threshold AND
                    # above logfoldchange threshold
                    dplyr::filter(
                        pvals_adj <= py$alpha,
                        abs(logfoldchanges) >= py$min_marker_l2fc)

            }) %>%
            bind_rows()

            # Add a column indicating the number of occurrences of the marker genes
            cluster.df <- df %>%
                group_by(names) %>%
                summarize(nclusters=n())

            df <- left_join(df, cluster.df, by='names') %>%
                unique()

            # Save to the list
            subset_marker_list[[condition]][[name]] <- df
        }

    }
}

```

## Tables (subset by `r py$cluster_col`) {.tabset}

Here, we compute marker genes between experimental groups within each cluster. 
Clusters are skipped if no marker genes were found.

```{r print_subset_markers, results='asis'}

# Remove data frames with zero marker genes
subset_marker_list <- lapply(
    subset_marker_list, function(list_i) {
        list_i <- list_i[sapply(list_i, nrow) > 0]
        return(list_i)
    }
)

for (condition in names(subset_marker_list)) {
    cat('###', condition, '{.tabset}\n\n')
    for (leiden in names(subset_marker_list[[condition]])) {
        cat('####', leiden, '{.tabset}\n\n')
        df <- subset_marker_list[[condition]][[leiden]]
        key <- paste0(
            "integrated_", 
            condition,
            "_in_cluster",
            leiden
        )

        cat('##### All markers {.tabset}\n\n')
        # Save total marker genes
        file.i <- file.path(py$output_dir,
                            paste0(key, '_all_markers.tsv'))
        write.table(df,
                    file.i,
                    quote=FALSE,
                    sep='\t',
                    row.names=FALSE,
                    col.names=TRUE)

        link.table(file.i)
        # Loop over to print 2000 top FDR genes in each cluster
        # Split the data frame by cluster
        df.list <- split(df, df$cluster)
        # Reorder the list
        df.list <- df.list[levels(py$adata$obs[[condition]])]
        # Remove NULL if present
        df.list <- df.list[!is.na(names(df.list))]
        for (name in names(df.list)) {
            cat('######', name, '\n\n')
            # Subset top 2000 FDR genes
            df <- df.list[[name]] %>%
                arrange(pvals_adj) %>%
                head(2000)
            # Remove rownames
            rownames(df) <- NULL
            subchunkify(paste0(key, '_', condition, '_allmarkers'))
            cat('\n\n')
        }


        # Clean the data frame to remove non-unique genes
        df <- do.call('rbind', df.list) %>%
            dplyr::filter(nclusters == 1)

        if (nrow(df) > 0) {

            cat('##### Unique markers {.tabset}\n\n')
            # Save total marker genes
            file.i <- file.path(py$output_dir,
                                paste0(key, '_unique_markers.tsv'))
            write.table(df,
                        file.i,
                        quote=FALSE,
                        sep='\t',
                        row.names=FALSE,
                        col.names=TRUE)

            link.table(file.i)

            # Loop over to print 2000 top FDR genes in each cluster
            # Split the data frame by cluster
            df.list <- split(df, df$cluster)
            # Reorder the list
            df.list <- df.list[levels(py$adata$obs[[condition]])]
            # Remove NULL if present
            df.list <- df.list[!is.na(names(df.list))]
            for (name in names(df.list)) {
                cat('######', name, '\n\n')
                # Subset top 2000 FDR genes
                df <- df.list[[name]] %>%
                    arrange(pvals_adj) %>%
                    head(2000)
                # Remove rownames
                rownames(df) <- NULL
                subchunkify(paste0(key, '_', name, '_uniquemarkers'))
                cat('\n\n')
            }
        }
    }
}

```


```{r save_adata}

# Assign file name to be saved
h5ad <- file.path(py$output_dir, "adata_integrated_markers.h5ad")

write_h5ad(
    py$adata,
    h5ad
    )
```

```{r prep_seurat}

# NOTE:
# We convert adata to seurat for easy manipulation in plotting steps by taking advantage of
# SeuratDisk package, as guided below:
# - https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html#converting-from-anndata-to-seurat-via-h5seurat
# - https://mojaveazure.github.io/seurat-disk/reference/Convert.html

# Assign filename for h5seurat
h5seurat <- file.path(py$output_dir, "adata_integrated_markers.h5seurat")

# Convert h5ad to h5seurat
SeuratDisk::Convert(
    h5ad,
    dest='h5seurat',
    overwrite=FALSE,
    verbose=FALSE
    )

# Import h5seurat
# NOTE: Metadata is omitted to avoid Error: Missing required datasets 'levels' and 'values'
# (see https://github.com/mojaveazure/seurat-disk/issues/109)
obj <- SeuratDisk::LoadH5Seurat(
    h5seurat,
    assays="RNA",
    meta.data=FALSE,
    misc=FALSE
    )

# Add metadata back to seurat obj
obj <- AddMetaData(obj, py$adata$obs)

# Copy normalized counts to RNA-data slot
obj@assays$RNA@data <- obj@assays$RNA@counts

# Extract and convert raw counts from dgRMatrix to dgCMatrix
raw.counts <- t(py$adata$layers['counts']) %>%
    as.matrix() %>%
    as("dgCMatrix")

# Add raw counts to RNA-counts slot
obj@assays$RNA@counts <- raw.counts

# Import sampletable
sample.df <- read.table(py$input_sampletable, header=TRUE, sep='\t')
sample.df$id <- 1:nrow(sample.df)

# Add a column storing cleaned barcodes to have the suffix matching samplename
# in the metadata
obj@meta.data <- obj@meta.data %>%
    left_join(sample.df[, c('samplename', 'id')], by='samplename')
obj$barcodes <- colnames(obj)
obj$barcodes <- sapply(strsplit(obj$barcodes, split='-'), function(x) x[1]) %>%
    unname()
obj$barcodes <- paste0(obj$barcodes, '-', obj$id)

# Replace barcodes with cleaned ones
obj <- RenameCells(obj, new.names=unname(obj$barcodes))

# Create a clean seurat obj
seurat <- CreateSeuratObject(
    counts=obj@assays[['RNA']]@counts,
    meta.data=obj@meta.data,
    assay="RNA"
    )

# Add normalized counts
seurat@assays[['RNA']]@data <- obj@assays[['RNA']]@data

# Add dim reductions
seurat@reductions[['umap']] <- obj@reductions[['umap']]
seurat@reductions[['unintegrated_umap']] <- obj@reductions[['unintegrated_umap']]
seurat@reductions[['pca']] <- obj@reductions[['pca']]


# Delete the h5seurat file
unlink(h5seurat)

# Delete unnecessary variables/objects
rm(raw.counts)
rm(obj)

```

```{r save_as_seurat}

# Save seurat-converted adata as seurat for further analysis
filename <- file.path(py$output_dir, "seurat_integrated_markers.rds")

saveRDS(
    seurat,
    filename
    )

```


## Genes of interest

We focus on genes of interest listed below.


```{r print_genes_of_interest}

for (name in names(py$genes)) {
    print(paste0(name, ": ", paste0(py$genes[[name]], collapse=", ")))
}

```


### Dotplot {.tabset}


```{r genes_of_interest_dotplot, results='asis'}

# Create a character vector storing genes of interest
genes_of_interest <- unique(unname(unlist(py$genes)))

for (key in py$marker_groups) {
    cat('####', key, '\n\n')
    # Create a dotplot using Seurat::DotPlot()
    p <- DotPlot(
        seurat,
        assay='RNA',
        features=genes_of_interest,
        group.by=key) +
        RotatedAxis()
    # Adjust width of the plot
    if (key %in% py$cluster_col_all) { height <- 16 } else { height <- 5 }
    # Print the plot
    subchunkify(paste0(key, '_dotplot'),
                input='plot', height=height, ggplotly=FALSE)
    cat('\n\n')
    # Save and link
    file.i <- file.path(py$output_dir,
                        paste0(key, '_genes_of_interest_dotplot.pdf')
                        )
    ggsave(file.i,
           ggplotify::as.ggplot(p),
           width=7,
           device="pdf",
           height=height)
    link.plot(file.i)
    cat('\n\n')
}


```

### Violin plot {.tabset}

```{r genes_of_interest_violin, results='asis'}

for (key in py$marker_groups) {
    cat('####', key, '{.tabset}\n\n')
    # Create a violin plot using Seurat::VlnPlot()
    for (gene in genes_of_interest) {
        cat('#####', gene, '\n\n')
        p <- VlnPlot(
            seurat,
            assay='RNA',
            features=gene,
            pt.size=0,
            group.by=key) +
            NoLegend() +
            theme(axis.text.x=element_text(size=8))

        # Adjust width of the plot
        if (key %in% py$cluster_col_all) { width <- 10 } else { width <- 7 }

        # Print the plot
        subchunkify(paste0(key, '_vlnplot'),
                    input='plot', width=width, ggplotly=TRUE)
        cat('\n\n')
        # Save and link
        file.i <- file.path(py$output_dir,
                            paste0(key, '_', gene, '_vlnplot.pdf')
                            )
        suppressMessages(
            ggsave(file.i,
                   ggplotify::as.ggplot(p),
                   width=width,
                   device="pdf",
                   )
            )
        link.plot(file.i)
        cat('\n\n')
    }
}
```

### Feature Plot {.tabset}

```{r genes_of_interest_featureplot, results='asis'}

for (key in py$marker_groups) {
    # Create a featureplot using Seurat::FeaturePlot()
    cat('#### UMAP by', key, '\n\n')
    lab <- ifelse(key == 'group', FALSE, TRUE)
    p <- DimPlot(seurat,
                 group.by=key,
                 pt.size=0.1,
                 reduction='umap',
                 label=lab,
                 repel=lab
                 )

    subchunkify(paste0(key, '_umap'),
                input='plot', 10, 8, ggplotly=FALSE)
    file.i <- file.path(py$output_dir,
                        paste0(key, '_umap.pdf')
                        )
    ggsave(file.i,
           p,
           width=10,
           height=8,
           device="pdf",
           )
    link.plot(file.i)
    cat('\n\n')
}

for (gene in genes_of_interest) {
    cat('####', gene, '\n\n')
    p <- FeaturePlot(
        seurat,
        features=gene,
        pt.size=0.1)
    # Print the plot
    subchunkify(paste0(key, '_', gene, '_featureplot'),
                input='plot', 10, 8, ggplotly=FALSE)
    cat('\n\n')
    # Save and link
    file.i <- file.path(py$output_dir,
                        paste0(key, '_', gene, '_featureplot.pdf')
                        )
    ggsave(file.i,
           p,
           width=10,
           height=8,
           device="pdf",
           )
    link.plot(file.i)
    cat('\n\n')
}
```


# Session info

```{r session_info, collapse=FALSE}
sessionInfo()
```
